### ISSUE TITLE
Request soon after premature close can result in database connection leak

### ISSUE THREAD

**Issue by matthew-white** [getodk/central-backend#485]
A client can cause a premature close. For example, if you request an OData feed using `curl`, then kill `curl` as the response is being received, then there may be a premature close. (The OData feed may have to be on the larger side for this to work.) A premature close logs an error, but by itself, it doesn't seem to cause an issue. For example, you can take the following steps:

- Run `select count(*) from pg_stat_activity where state <> 'idle'`. The count should be 1 (corresponding to this query itself).
- Request an OData feed and cause a premature close.
- Immediately run the count query again. The count should be 2.
- Wait a few seconds, then run the query again. The count should be back to 1.

However, if something (either a request or a worker) attempts to acquire a database connection soon after the premature close, things can go wrong. The OData feed uses a Slonik stream, and it looks like if the connection pool gives out the same connection that the stream was using, that becomes an issue. To see this, specify the Slonik `interceptors` option to `createPool()`:

```js
interceptors: [{
  afterPoolConnection: ({ connectionId }) => {
    console.log('using connection', connectionId);
  },
  beforePoolConnectionRelease: ({ connectionId }) => {
    console.log('releasing connection', connectionId);
  }
}]
```

Now take the following steps:

- Request an OData feed and cause a premature close.
- Immediately after, send another request. (You may find it easiest to send the same request, but any request should reproduce.)
- If the stream's connection has been given out, the request will not receive a response.

You may have to try a few times to see this behavior. You'll see logs that look like this:

```
using connection HSEiX93XRJK1gDyV70K_Cg-22
releasing connection HSEiX93XRJK1gDyV70K_Cg-22
using connection HSEiX93XRJK1gDyV70K_Cg-22
releasing connection HSEiX93XRJK1gDyV70K_Cg-22
using connection HSEiX93XRJK1gDyV70K_Cg-22
releasing connection HSEiX93XRJK1gDyV70K_Cg-22
using connection HSEiX93XRJK1gDyV70K_Cg-22
releasing connection HSEiX93XRJK1gDyV70K_Cg-22
using connection HSEiX93XRJK1gDyV70K_Cg-27
using connection HSEiX93XRJK1gDyV70K_Cg-29
releasing connection HSEiX93XRJK1gDyV70K_Cg-29
::1 - - [14/May/2022:14:51:29 +0000] "GET /v1/projects/1/forms/foo.svc/Submissions HTTP/1.1" 200 -
releasing connection HSEiX93XRJK1gDyV70K_Cg-27
Error [ERR_STREAM_PREMATURE_CLOSE]: Premature close
    at ServerResponse.onclose (internal/streams/end-of-stream.js:121:38)
    at ServerResponse.emit (events.js:412:35)
    at ServerResponse.emit (domain.js:470:12)
    at Socket.onServerResponseClose (_http_server.js:221:44)
    at Socket.emit (events.js:412:35)
    at Socket.emit (domain.js:470:12)
    at TCP.<anonymous> (net.js:675:12)
using connection HSEiX93XRJK1gDyV70K_Cg-27
```

In the logs above, `HSEiX93XRJK1gDyV70K_Cg-27` is the ID of the connection that `stream()` used. The connection is given out to the next request, which seems to be unable to use it.

You may have to try the steps above several times before reproducing this behavior. Once you do, run the count query from above. The count is now 2 and as far as I can tell, never goes back down, indicating a connection leak. Even if you kill the second request, the count remains at 2.

If you repeat this process enough times such that the count reaches 11, the connection pool will be full. At that point, Problem responses will be returned with the message `Completely unhandled exception: timeout exceeded when trying to connect`.

I believe that this is an upstream issue and have filed an issue with Slonik.

(By the way, I think that the latest version of Slonik may pass slightly different information to the interceptor methods, so I'm not sure that the interceptor above would work as-is in the latest version.)

**Comment by lognaturel** [comment]
https://github.com/gajus/slonik/issues/347

**Comment by matthew-white** [comment]
Closed by #604.
